src/module-kount-control/Model/CustomerLogin.php

```diff
@@ -0,0 +1,117 @@
+        \Swarming\KountControl\Model\ControlApi\Event $eventService,
+        \Swarming\KountControl\Model\ControlApi\TrustedDevice $trustedDeviceService,
+        \Swarming\KountControl\Helper\Config $kountControlConfig
+    ) {
+        $this->customerSession = $customerSession;
+        $this->kountConfig = $kountConfig;
+        $this->loginService = $loginService;
+        $this->eventService = $eventService;
+        $this->trustedDeviceService = $trustedDeviceService;
+        $this->kountControlConfig = $kountControlConfig;
+    }
+
+    /**
+     * Goes through KountControl diagram and initiate API requests to KountControl
+     *
+     * @param $sessionId
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Swarming\KountControl\Exception\ConfigException
+     * @throws \Swarming\KountControl\Exception\NegativeApiResponse
+     * @throws \Swarming\KountControl\Exception\ParamsException
+     * @throws \Swarming\KountControl\Exception\PositiveApiResponse
+     */
+    public function login($sessionId)
+    {
+        if (!$this->kountControlConfig->isLoginServiceEnabled()) {
+            throw new \Swarming\KountControl\Exception\ConfigException(__('KountControl: Login service disable'));
+        }
+
+        $userId = $this->customerSession->getCustomerId();
+        $clientId = $this->kountConfig->getMerchantNumber();
+        if ($sessionId === '' || $userId === '' || $this->kountConfig->getMerchantNumber() === '') {
+            throw new \Swarming\KountControl\Exception\ParamsException(__('KountControl: lost POST params. '
+                . '$sessionId = "%1"; $userId = "%2"; $clientId', $sessionId, $userId, $clientId));
+        }
+
+        $loginResult = $this->loginService->executeApiRequest($sessionId, $clientId);
+        $this->eventService->setLoginResult($loginResult);
+        if ($loginResult['decision'] === \Swarming\KountControl\Model\ControlApi\Event::CHALLENGE_DECISION) {
+            if (!isset($loginResult['deviceId'])) {
+                throw new \Swarming\KountControl\Exception\ParamsException(__(
+                    'KountControl: lost POST params. API login decision is "%1". $deviceId is not set.',
+                    $loginResult['decision']
+                ));
+            }
+            $this->eventService->executeApiRequest($sessionId, $clientId);
+            if ($this->kountControlConfig->isTrustedDeviceEnabled()) {
+                $this->trustedDeviceService->setDeviceId($loginResult['deviceId']);
+                $this->trustedDeviceService->executeApiRequest($sessionId, $clientId);
+            }
+            throw new \Swarming\KountControl\Exception\PositiveApiResponse(__(
+                'KountControl: API Login decision is "%1"',
+                $loginResult['decision']
+            ));
+        } else {
+            $this->eventService->executeApiRequest($sessionId, $clientId);
+            throw new \Swarming\KountControl\Exception\NegativeApiResponse(__(
```

Comment:
Should this be thrown every time you dont meet a challenge decision?

Reply:
Yes. In this place can be only "Block" decision. And the further logic waits this type of exception for logout customer and other actions.